Use types are origins for incremental KAPT and track generated source

gavra0 opened the pull request:
This change introduces tracking of generated sources structure in order
to e.g track classpath changes impacting generated sources. This fixes KT-42182.

Also, origin tracking for isolating processors is now using types, allowing
for origin elements from classpath. This fixes KT-34340. However, classpath
origin is used only to invalidate generated files when the type changes and
processing will not be requested for that type. This is in line with the
incap spec.

max-kammerer commented on libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/KaptIncrementalWithAggregatingApt.kt:
```diff
@@ -275,8 +275,9 @@ class KaptIncrementalWithAggregatingApt : KaptIncrementalIT() {
-                ), getProcessedSources(output)
+                    fileInWorkingDir("build/tmp/kapt3/stubs/main/error/NonExistentClass.java").canonicalPath,
+                    fileInWorkingDir("build/generated/source/kapt/main/bar/WithAnnotationGenerated.java").canonicalPath.takeUnless { isBinary },
+                ).filterNotNull().toSet(), getProcessedSources(output)
```
You can use `setOfNotNull()`

gavra0 replied:
Tests are compiled with apiVersion 1.3 and `setOfNotNull` got added in 1.4.

max-kammerer commented on libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/KaptIncrementalWithIsolatingApt.kt:
```diff
@@ -192,13 +194,94 @@ class KaptIncrementalWithIsolatingApt : KaptIncrementalIT() {
+        classpathTypeSource.writeText(classpathTypeSource.readText().replace("}", "int i = 10;\n}"))
+        project.build("build") {
+            assertSuccessful()
+            assertEquals(
```
Is first build affect same files? Is there any sense to check them in first build?

max-kammerer commented on plugins/kapt3/kapt3-base/test/org/jetbrains/kotlin/kapt3/base/incremental/JavaClassCacheManagerTest.kt:
```diff
@@ -155,38 +157,14 @@ class JavaClassCacheManagerTest {
     }
 
-    @Test
-    fun testWithAnnotations() {
```
It's not clear why this test was deleted

gavra0 replied:
`invalidateEntriesAnnotatedWith` function is removed and we are now relying on `IncrementalAptCache.aggregatedTypes` to find all types annotated by aggregating annoattions.

max-kammerer commented:
There are several failing test on TC:
```


KotlinKaptContextTestGenerated.testOverloads


KotlinKaptContextTestGenerated.testNestedClasses


KotlinKaptContextTestGenerated.testSimple


IrKotlinKaptContextTestGenerated.testOverloads


IrKotlinKaptContextTestGenerated.testNestedClasses


IrKotlinKaptContextTestGenerated.testSimple


```

max-kammerer commented on plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/incrementalProcessors.kt:
```diff
@@ -70,11 +70,17 @@ class IncrementalProcessor(private val processor: Processor, private val kind: D
-    fun getGeneratedToSources() = dependencyCollector.value.getGeneratedToSources()
-    fun getGeneratedToSourcesAll() = dependencyCollector.value.getGeneratedToSourcesAll()
+
+    /** Mapping fromm generated file to type that were used as originating elements. For aggregating APs types will be [null]. */
```
Misspelling: from

max-kammerer commented on plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/IncrementalAptCache.kt:
```diff
@@ -10,16 +10,19 @@ import java.io.Serializable
-    fun updateCache(processors: List<IncrementalProcessor>): Boolean {
+    fun updateCache(processors: List<IncrementalProcessor>, failedToAnalyzeSources: Boolean): Boolean {
+        if (failedToAnalyzeSources) {
+            invalidateCache()
```
Will it imply that next execution would be full rebuild and there wouldn't be any attempt to perform IC?

gavra0 commented:
I'm looking into failing tests.

gavra0 commented:
I've addressed comments in `f7a2521` and in fixed failing tests in `25e3348`. Also, I had to rebase to avoid conflict but there are no changes in the first 3 commits.

max-kammerer commented on plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/annotationProcessing.kt:
```diff
@@ -40,6 +41,13 @@ fun KaptContext.doAnnotationProcessing(
 
     val compilerAfterAP: JavaCompiler
     try {
+        if (javaSourceFiles.isEmpty() && aggregatedTypes.isEmpty() && additionalSources.isEmpty()) {
```
`aggregatedTypes.isEmpty()` a bit strange for me, do it make sense on classpath changes in case when aggregation processor presented? (Then could we test such case?)

max-kammerer replied:
Or if same case is covered by `additionalSources.isEmpty` check could first one be removed?

gavra0 replied:
`additionalSources` is only used in tests, so I need all 3 checks to be present.

I'll rename `aggregatedTypes` to `typesToProcess` as `doAnnotationProcessing` should not be aware of isolating/aggregating APs. It should just know that there are some types to process.

It is possible for `aggregatedTypes` to contain some elements, while the sources are empty. This is in case that isolating AP generated `AGen.class`, which contains aggregating annotation that generated `AGen2.java`, and in `AGen2.java` a classpath type `X` is mentioned. If in the second run `X` changes, we need to reprocess `AGen.class`.

That being said, I'll add a test case to cover this.

max-kammerer replied:
It's still not clear why it should be reprocessed by APs/kapt, as I understand any change in X shouldn't affect annotation processor, but just AGen2.java recompilation

max-kammerer replied:
or it could be in case when AP itself tries to inspect X internals?

gavra0 replied:
Case in which `X` internals are inspected is precisely why `T` is reprocessed. E.g for an AP under incremental rules, it is possible to get type `T` which is annotated with a claimed annotation, and then to explore types visible from `T` and to make decisions based on that. The only limitation is that we assume if `T` uses `X` and `X` uses `Y`, `Y` needs to be in ABI of `X` in order to impact `T`. This is how Gradle handles it, and we are doing the same in kapt.

max-kammerer commented on plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/cache.kt:
```diff
@@ -19,8 +19,18 @@ class JavaClassCacheManager(val file: File) : Closeable {
+                val typeInformation = SourceFileStructure(classFile.toURI()).also {
+                    it.addDeclaredType(type)
+                }
+                javaCache.addSourceStructure(typeInformation)
```
NB: `JavaClassCache.invalidateTypesForFiles ` is unused now

When class files entries are deleted from cache when AP origins are not presented any more in sources

max-kammerer replied:
Looks like it's `javaCache.invalidateDataForTypes(impactedTypes)`

gavra0 replied:
I'll remove `invalidateTypesForFiles`. Correct `invalidateDataForTypes` is used to remove `.class` (and `.java`) file type information.

max-kammerer commented on plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/cache.kt:
```diff
@@ -39,38 +49,70 @@ class JavaClassCacheManager(val file: File) : Closeable {
+            javaCache.invalidateDataForTypes(impactedTypes)
+            aptCache.invalidateAggregating()
+            // for isolating, invalidate both own types and classpath types
+            aptCache.invalidateIsolatingForOriginTypes(impactedTypes)
```
Maybe `aptCache.invalidateIsolatingForOriginTypes(impactedTypes + dirtyClasspathFqNames) better?

gavra0 replied:
I wanted to avoid creating additional set, but considering we iterate over all isolating mappings, it is more efficient to do so.

max-kammerer replied:
ok

max-kammerer commented on plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/cache.kt:
```diff
@@ -39,38 +49,70 @@ class JavaClassCacheManager(val file: File) : Closeable {
+        val impactedTypes = javaCache.getAllImpactedTypes(changes)
+        // check isolating with origins from the classpath
+        impactedTypes.addAll(aptCache.getIsolatingGeneratedTypesForOrigins(changes.dirtyFqNamesFromClasspath, javaCache::getTypesForFiles))
+        if (changes.sourceChanges.isNotEmpty() || impactedTypes.isNotEmpty()) {
```
Could classpath change (e.g. lib dependency added) add additional origins for aggregated AP? How would work kapt in such case?

max-kammerer replied:
Is classpath change affects only compilation of generated classes or can add additional origins for AP or such origins obtained only within module?

gavra0 replied:
Aggregating APs can have origins only within module.

For isolating, it is possible to have origin outside of module as APs are able to specify any originating element. It the origin type from classpath changes, any generated files from it will be removed. However, KAPT will not reprocess that type. It is expected that APs will follow the same type resolution path as they initially did in order to re-generate them. Also, at least one source from module should have type dependency on the classpath type, otherwise it would not be possible for isolating AP to reach it. Supporting these origins is just to make it easier on the isolating APs (see KT-34340).

In general, there are 2 stages in determining impacted types:
a) following type dependencies
b) following annotation processing dependencies

For a), any mention of type `X` in `T` will cause `T` to become invalid if `X` changes. This becomes transitive if `X` is part of `T`s ABI.
For b), if any source/type needs to be reprocessed, we must reprocess all aggregating annotations (as processing any sources/class could generated sources that contain aggregating annotation). Also, all isolating generated types are invalidated if their origins have changed (this is transitive).

gavra0 commented:
Commit 50270c1 addresses the last set of comments and it adds test to check that kapt behaves correctly if there are only types to reprocess (i.e no .java sources to reprocess).

max-kammerer commented on libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/resources/testProject/kaptIncrementalWithParceler/mylibrary/src/main/java/com/example/lib/ExampleParcel.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright 2011-2015 John Ericksen
```
Could you just mention project from where sample was obtained and remove copyright?
Otherwise there are another TC fail:
```
  junit.framework.AssertionFailedError: The following files contain third-party copyrights and no license information. Please update license/README.md accordingly:
  libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/resources/testProject/kaptIncrementalWithParceler/mylibrary/src/main/java/com/example/lib/ExampleParcel.java:  * Copyright 2011-2015 John Ericksen
  junit.framework.AssertionFailedError: The following files contain third-party copyrights and no license information. Please update license/README.md accordingly:
  libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/resources/testProject/kaptIncrementalWithParceler/mylibrary/src/main/java/com/example/lib/ExampleParcel.java:  * Copyright 2011-2015 John Ericksen
```

gavra0 replied:
Done in  583dae2.

max-kammerer commented on plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/annotationProcessing.kt:
```diff
@@ -32,14 +32,21 @@ fun KaptContext.doAnnotationProcessing(
     processors: List<IncrementalProcessor>,
     additionalSources: JavacList<JCTree.JCCompilationUnit> = JavacList.nil(),
-    aggregatedTypes: List<String> = emptyList()
+    typesToReprocess: List<String> = emptyList()
```
Actually it should be types for binary classes, maybe use  `binaryTypesToReprocess` then (not sure that it's the best one)

gavra0 replied:
Done

max-kammerer commented:
Many thanks for your work! 
PR is merged manually
